library/dhm.c

```diff
@@ -316,19 +316,18 @@ int mbedtls_dhm_make_public( mbedtls_dhm_context *ctx, int x_size,
 
     if( mbedtls_mpi_cmp_int( &ctx->P, 0 ) == 0 )
         return( MBEDTLS_ERR_DHM_BAD_INPUT_DATA );
+    n_bits = mbedtls_mpi_bitlen( &ctx->P );
 
     /*
      * generate X and calculate GX = G^X mod P
      */
     do
     {
-        MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
-
-        while( mbedtls_mpi_cmp_mpi( &ctx->X, &ctx->P ) >= 0 )
-            MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &ctx->X, 1 ) );
-
-        if( count++ > 10 )
+        if( count++ > 30 )
             return( MBEDTLS_ERR_DHM_MAKE_PUBLIC_FAILED );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &ctx->X, 8 * x_size - n_bits ) );
```

Comment:
Even in an LTS branch, I think unifying the code of `mbedtls_dhm_make_params` and `mbedtls_dhm_make_public`, which differs only by the choice of error code, would be better.